kernel: early_random: Internal early 'random' subsys fixes

ceolin opened the pull request:
- Rename `z_early_boot_rand_get` with `z_early_rand_get` to get consistent with other early functions.
- Make z_early_rand_get a weak symbol
   - Allow targets come up with their own early random generator since the default can be NOT so random due constraints.
 - Fixes for z_early_rand_get

    - The early random get function was making many wrong assumptions
    about random subsys and entropy drivers. First, it was assuming
    that entropy_get_entropy() would be ISR safe, that is not right,
    the driver has an ISR safe callback and if it is not implemented
    or not working it is not ok using the other callback.
    Second, the fallback to the random subsys is even more problematic
    since they can use kernel services to protect internal states and be
    thread-safe.

    - Another incorrect thing in this function was the guard around it.
    It was needed by features like stack randomization and stack canaries,
    and not when those conditions were match. Just remove it and in case
    it is not needed the linker will take care of it.

    - The drawback of this change is that in the absence of an entropy
    generator with support to be called from ISR the randomness is very
    weak.

Commit:
kernel: random: Rename early random get function

Rename z_early_boot_rand_get with z_early_rand_get to get consistent
with other early functions.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: random: Make z_early_rand_get a weak symbol

Allow targets come up with their own early random generator
since the default can be NOT so random due constraints.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: Fixes for z_early_rand_get

The early random get function was making many wrong assumptions
about random subsys and entropy drivers. First, it was assuming
that entropy_get_entropy() would be ISR safe, that is not right,
the driver has an ISR safe callback and if it is not implemented
or not working it is not ok using the other callback.
Second, the fallback to the random subsys is even more problematic
since they can use kernel services to protect internal states and be
thread-safe.

Another incorrect thing in this function was the guard around it.
It was needed by features like stack randomization and stack canaries,
and not when those conditions were match. Just remove it and in case
it is not needed the linker will take care of it.

The drawback of this change is that in the absence of an entropy
generator with support to be called from ISR the randomness is very
weak.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

peter-mitsis commented on kernel/init.c:
```diff
@@ -508,46 +508,38 @@ static FUNC_NORETURN void switch_to_main_thread(char *stack_ptr)
 {
-#ifdef CONFIG_ENTROPY_HAS_DRIVER
-	const struct device *const entropy = DEVICE_DT_GET_OR_NULL(DT_CHOSEN(zephyr_entropy));
+	static uint64_t state = 123456789UL;
```
Two minor items about this constant 123456789UL.

The first is that since _state_ is of type _uint64_t_, it makes me think that we should be using the _ULL_ suffix instead of _UL_.

The second is that I am wary about this being a hard-coded constant for the fallback case. I am imagining a scenario where there is a fleet of devices/products so predictable in their timing such that when their software invokes `z_early_rand_get()` that they all get the same results--thereby negating some of the randomness. To cover that case, we could make this _123456789_ constant configurable. Thoughts?

ceolin replied:
Yep, I thought about it. This constant I got from the timer based random implementation, just  tried to get rid of existent issues. I also have made it as weak symbol to allow applications to come up with a better random generator if possible. Make the constant will still be very predictable but at least won't be an universal value.

peter-mitsis replied:
I forgot all about the weak attribute when I wrote that. That totally allows for any necessary customisation.

ceolin replied:
@peter-mitsis even allowing replace this function, I think it is worth to add a symbol to customize it. Will send a new version doing it.

Commit:
random: kconfig: Allow customizing timer initial state

Add a build option to allow changing the initial state used
in the timer based random generator and by the kernel in the
early random number generator.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: init: Build constant in early random generator

Use a Kconfig symbol for initial state in the early
random number generator.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
random: timer: Use build constant for seed

Use Kconfig symbol for initial seed.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@peter-mitsis Just moved the seed to Kconfig

ceolin commented:
@jhedberg @fabiobaltieri just to be clear, these pr contains bug fixes, it does not have an issue associated with because I found them when working in something else.

fabiobaltieri commented:
ok I changed back the milestone

@jhedberg can you take another look at this? Was there something in particular that made you remove the milestone before?

@andyross could you review the `kernel/` parts? Would be nice to have approval before cutting RC3

nashif commented:
we all love random fixes... :-)

jhedberg commented:
@andyross could you please review this one as well? Thanks.
